protocols/gossipsub/src/behaviour.rs

```diff
@@ -1757,6 +1754,12 @@ where
             self.mcache.observe_duplicate(&msg_id, propagation_source);
             return;
         }
+
```

Comment:
why did we move the sending of IDONTWANT messages down here? We should send them before validation, [as the spec states](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.2.md#basic-scenario)

Reply:
I have moved the sending logic above the duplicate validation, although the spec mentions that idontwant can be sent without validating the messages seems like it a possible case: we receive an invalid message and `idontwant` is sent for the same. 

again I may be wrong on my assumptions

Reply:
this is something that has been brought in the past. as the [doc states](https://docs.rs/libp2p/latest/libp2p/gossipsub/struct.ConfigBuilder.html#method.message_id_fn):
> One example is content based addressing, where this function may be set to hash(message). This would prevent messages of the same content from being duplicated.

So one should set the message id to be a collision resistant hash function.

Reply:
oh that makes sense. thanks for the clarification, I'll make the required change.